daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -642,94 +643,171 @@ private[lf] object Speedy {
+                  case V.ValueList(ls) =>
+                    SValue.SList(ls.map(go(elemType, _)))
+
+                  // textMap
+                  case V.ValueTextMap(entries) =>
+                    SValue.SGenMap(
+                      isTextMap = true,
+                      entries = entries.iterator.map { case (k, v) =>
+                        SValue.SText(k) -> go(elemType, v)
+                      },
+                    )
                   case _ =>
-                    crash(s"definition for variant $id not found")
+                    typeMismatch
                 }
-              case None =>
-                throw SpeedyHungry(
-                  SResultNeedPackage(
-                    id.packageId,
-                    pkg => {
-                      compiledPackages = pkg
-                      returnValue = go(value)
-                    },
-                  )
-                )
+              case keyType :: valueType :: Nil =>
+                value match {
+                  // genMap
+                  case V.ValueGenMap(entries) =>
+                    SValue.SGenMap(
+                      isTextMap = false,
+                      entries = entries.iterator.map { case (k, v) =>
+                        go(keyType, k) -> go(valueType, v)
+                      },
+                    )
+                  case _ =>
+                    typeMismatch
+                }
+              case _ =>
+                typeMismatch
+            }
+          case TTyCon(tyCon) =>
+            val signature = compiledPackages.signatures(tyCon.packageId)
+            value match {
+              case V.ValueRecord(_, fields) =>
+                signature.lookupRecord(tyCon.qualifiedName) match {
+                  case Right((params, DataRecord(fieldsDef))) =>
+                    lazy val subst = (params.toSeq.view.map(_._1) zip tyArgs).toMap
+                    val n = fields.length
+                    var i = 0
+                    val values = new util.ArrayList[SValue](n)
+                    while (i < n) {
+                      val typDef = fieldsDef(i)._2
+                      val typ = if (tyArgs.isEmpty) {
+                        // optimization
+                        typDef
+                      } else {
+                        replaceParameters(subst, typDef)
+                      }
+                      values.add(go(typ, fields(i)._2))
+                      i += 1
+                    }
+                    SRecord(tyCon, fieldsDef.map(_._1), values)
+                  case Left(err) => crash(err)
+                }
+              case V.ValueVariant(_, constructor, value) =>
+                signature.lookupVariant(tyCon.qualifiedName) match {
+                  case Right((params, variantDef)) =>
+                    val rank = variantDef.constructorRank(constructor)
+                    val typDef = variantDef.variants(rank)._2
+                    val typ = if (tyArgs.isEmpty) {
```

Comment:
proving my point that doing it once in the callee might be a good idea :slightly_smiling_face: